asmcomp/debug/asm_directives.ml

```diff
@@ -0,0 +1,1004 @@
+    in
+    size size_of_symbol (Sub (This, Symbol symbol))
+  | _ -> ()
+
+let label label_name = const_machine_width (Label label_name)
+
+let define_label label_name =
+  let typ : Directive.thing_after_label =
+    if current_section_is_text () then Code
+    else Machine_width_data
+  in
+  emit (New_label (string_of_label label_name, typ))
+
+let sections_seen = ref []
+
+(* Modified version of [Target_system.system] for easier matching in
+   [switch_to_section], below. *)
+type derived_system =
+  | Linux
+  | MinGW_32
+  | MinGW_64
+  | Win32
+  | Win64
+  | Cygwin
+  | MacOS_like
+  | FreeBSD
+  | NetBSD
+  | OpenBSD
+  | Generic_BSD
+  | Solaris
+  | Dragonfly
+  | GNU
+  | BeOS
+  | Unknown
+
+let derived_system () =
+  match TS.system (), TS.machine_width () with
+  | Linux, _ -> Linux
+  | Windows Cygwin, _ -> Cygwin
+  | Windows MinGW, Thirty_two -> MinGW_32
+  | Windows MinGW, Sixty_four -> MinGW_64
+  | Windows Native, Thirty_two -> Win32
+  | Windows Native, Sixty_four -> Win64
+  | MacOS_like, _ -> MacOS_like
+  | FreeBSD, _ -> FreeBSD
+  | NetBSD, _ -> NetBSD
+  | OpenBSD, _ -> OpenBSD
+  | Generic_BSD, _ -> Generic_BSD
+  | Solaris, _ -> Solaris
+  | Dragonfly, _ -> Dragonfly
+  | GNU, _ -> GNU
+  | BeOS, _ -> BeOS
+  | Unknown, _ -> Unknown
+
+let switch_to_section (section : section) =
+  let first_occurrence =
+    if List.mem section !sections_seen then false
+    else begin
+      sections_seen := section::!sections_seen;
+      true
+    end
+  in
+  current_section := Some section;
+  let names, flags, args =
+    let text () = [".text"], None, [] in
+    let data () = [".data"], None, [] in
+    let rodata () = [".rodata"], None, [] in
+    let system = derived_system () in
+    match section, TS.architecture (), system with
+    | Text, _, _ -> text ()
+    | Data, _, _ -> data ()
+    | DWARF dwarf, _, MacOS_like ->
+      let name =
+        match dwarf with
+        | Debug_info -> "__debug_info"
+        | Debug_abbrev -> "__debug_abbrev"
+        | Debug_aranges -> "__debug_aranges"
+        | Debug_loc -> "__debug_loc"
+        | Debug_str -> "__debug_str"
+        | Debug_line -> "__debug_line"
+      in
+      ["__DWARF"; name], None, ["regular"; "debug"]
+    | DWARF dwarf, _, _ ->
+      let name =
+        match dwarf with
+        | Debug_info -> ".debug_info"
+        | Debug_abbrev -> ".debug_abbrev"
+        | Debug_aranges -> ".debug_aranges"
+        | Debug_loc -> ".debug_loc"
+        | Debug_str -> ".debug_str"
+        | Debug_line -> ".debug_line"
+      in
+      let flags =
+        if first_occurrence then
+          Some ""
+        else
+          None
+      in
+      let args =
+        if first_occurrence then
+          ["%progbits"]
+        else
+          []
+      in
+      [name], flags, args
+    | (Eight_byte_literals | Sixteen_byte_literals), (ARM | AArch64 | Z), _
+    | (Eight_byte_literals | Sixteen_byte_literals), _, Solaris ->
+      rodata ()
```

Comment:
Wouldn't it be better if the architectures exposed some constant to configure that ? It seems like it's going to be horrible to track that this needs to be updated in case a new architecture (for instance RiscV) is introduced.